File: src/api/dotnet/StringSymbol.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    StringSymbol.cs

Abstract:

    Z3 Managed API: String Symbols

Author:

    Christoph Wintersteiger (cwinter) 2012-11-23

Notes:
    
--*/

using System;
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace Microsoft.Z3
{

    /// <summary>
    /// Named symbols
    /// </summary>
    public class StringSymbol : Symbol
    {
        /// <summary>
        /// The string value of the symbol.
        /// </summary>
        /// <remarks>Throws an exception if the symbol is not of string kind.</remarks>
        public string String
        {
            get
            {

                if (!IsStringSymbol())
                    throw new Z3Exception("String requested from non-String symbol");
                return Native.Z3_get_symbol_string(Context.nCtx, NativeObject);
            }
        }

        #region Internal
        internal StringSymbol(Context ctx, IntPtr obj)
            : base(ctx, obj)
        {
            Debug.Assert(ctx != null);
        }

        internal StringSymbol(Context ctx, string s)
            : base(ctx, Native.Z3_mk_string_symbol(ctx.nCtx, s))
        {
            Debug.Assert(ctx != null);
        }

#if DEBUG
        internal override void CheckNativeObject(IntPtr obj)
        {
            if ((Z3_symbol_kind)Native.Z3_get_symbol_kind(Context.nCtx, obj) != Z3_symbol_kind.Z3_STRING_SYMBOL)
                throw new Z3Exception("Symbol is not of string kind");

            base.CheckNativeObject(obj);
        }
#endif
        #endregion
    }
}

File: src/api/java/StringSymbol.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    StringSymbol.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/ 

package com.microsoft.z3;

import com.microsoft.z3.enumerations.Z3_symbol_kind;

/**
 * Named symbols
 **/
public class StringSymbol extends Symbol
{
    /**
     * The string value of the symbol.
     * Remarks: Throws an exception if the
     * symbol is not of string kind.
     **/
    public String getString()
    {
        return Native.getSymbolString(getContext().nCtx(), getNativeObject());
    }

    StringSymbol(Context ctx, long obj)
    {
        super(ctx, obj);
    }

    StringSymbol(Context ctx, String s)
    {
        super(ctx, Native.mkStringSymbol(ctx.nCtx(), s));
    }

    @Override
    void checkNativeObject(long obj)
    {
        if (Native.getSymbolKind(getContext().nCtx(), obj) != Z3_symbol_kind.Z3_STRING_SYMBOL
                .toInt()) {
            throw new Z3Exception("Symbol is not of String kind");
        }
        super.checkNativeObject(obj);
    }
}
